File: server/Convert.py
def serializeFrame(frame, log):
  arr = []
  for (column = 0; column < 8; column += 1)
    for (panel = Ypix / 8 - 1; panel >= 0; panel -= 1)
      if (panel % 2 == 1) #TODO do dis better
        for (block = 0; block < 6; block += 1)
          for (y = 0; y < 8; y += 1)
            x = (7 - column) + (block * 8)
            arr.append(frame[y + (panel * 8)][x])

      else
        for (block = 0; block < 6; block += 1)
          for (y = 7; y >= 0; y -= 1)
            x = Xpix - 1 - (7 - column) - (block * 8)
            arr.append(frame[y + (panel * 8)][x])

      for (i = 0; i < 8; i += 1)
        if (i == column) arr.append(0)
        else arr.append(1)

  newArr = []
  for (i = 0; i < len(arr); i += 8)
    byte = ''
    for (bit = 0; bit < 8; bit += 1)
      byte += str(arr[bit + i])

    byte = int(byte, 2)
    newArr.append(byte)

  arr = newArr

  if (log) print(str(arr))
  return arr

def serializeFrames(frames):
  newFrames = []
  for frame in frames:
    newFrames.append(serializeFrame(frame))
  return newFrames

File: server/convert.js
const Ypix = 72
const Xpix = 48
const serializeFrame = require('./serializeFrame.js')(Xpix, Ypix)
const uploadDuration = 250
const frameDB = require('./frameDB.js')
const isPi = require('detect-rpi')()
const writeToPython = isPi ? require('./writeToPython.js') : async () => { }
const getTimeFrame = require('./clock/getTimeFrame.js')

let writeTimeTimeout
async function writeTime(loop = false, timestamp = Date.now()) {
	if (writeTimeTimeout) clearTimeout(writeTimeTimeout)
	const frame = getTimeFrame(timestamp)
	const bytes = serializeFrame(frame)
	await writeToPython(bytes)
	if (loop) writeTimeTimeout = setTimeout(() => writeTime(loop), Math.max(2000, uploadDuration))
}

async function stopWriteTime(loadStandardFrame) {
	if (writeTimeTimeout) clearTimeout(writeTimeTimeout)
	if (loadStandardFrame) await loadAnimation({ id: 'standard' })
}

let animationTimeout
async function loadAnimation({ id, frames, serializedFrames, interval = 0, private = false, framePos = 0 }) {
	await stopWriteTime(false)
	// requires either id, frames, or serializedFrames
	// case id: 1. retrieve animation from db 2. serialize 3. upload
	// case frames 1. serialize 2. upload
	// case serializedFra,es 1. upload

	clearTimeout(animationTimeout)
	let animation
	if (id) {
		animation = await frameDB.getAnimation(id, private)
		frames = animation.frames
		interval = animation.interval
	}
	if (frames) {
		serializedFrames = frames.map((frame) => serializeFrame(frame))
	}

	await writeToPython(serializedFrames[framePos])

	if (serializedFrames.length > 1) {
		if (++framePos > serializedFrames.length - 1) framePos = 0
		animationTimeout = setTimeout(() => {
			loadAnimation({ serializedFrames, interval, framePos })
		}, Math.max(0, interval - uploadDuration / 2))
	}

	return animation
}

module.exports = {
	frameDB,
	loadAnimation,
	writeTime,
	stopWriteTime,
}